zk/src/org/zkoss/zk/ui/impl/DesktopImpl.java

```diff
@@ -167,7 +170,19 @@ public class DesktopImpl implements Desktop, DesktopCtrl, java.io.Serializable {
 	//don't create it dynamically because PageImp._ip bind it at constructor
 	private transient Execution _exec;
 	/** A list of ScheduleInfo; must be thread safe */
-	private final ConcurrentLinkedQueue<ScheduleInfo<? extends Event>> _schedInfos = new ConcurrentLinkedQueue<ScheduleInfo<? extends Event>>();
+
+	// ZK-4194
+	private final AtomicInteger _scheduleInfoReadCount = new AtomicInteger(0);
+
+	// ZK-4194
+	private final AtomicInteger _scheduleInfoWriteCount = new AtomicInteger(0);
+
+	// ZK-4194, CacheBuilder doesn't support order for a cache, so we use Read/Write Count to simulate the order.
```

Comment:
This cache is to avoid a disconnected desktop (no connected, corresponding browser tab) from stacking ScheduleInfo sent by a server push EventQueue.
Cache is like a Map which is not an ordered collection, we use read/write count to serve as an index.

Reply:
this will break if someone insert a new line of code between these lines, so I prefer to keep the info. on each variable.